document `Tree` component (#1639)
iTwin/stratakit: apps/website/src/content/docs/components/stratakit/Tree.md

The **Tree** component exposes application data as a hierarchy of nested, expandable levels. **Trees** can be explored and organized in different ways. They may offer searching, filtering, and sorting functionality. 

Like the [**Accordion**](/components/accordion), the **Tree** is useful for progressively disclosing content. Unlike the **Accordion**, this content can be nested to multiple levels of depth.

| Use case                                                     | [Tree](/components/tree)                     | [Accordion](/components/accordion) | 
| ------------------------------------------------------------ | -------------------------------------------- | ---------------------------------- | 
| Progressive disclosure of content (several levels of data)   | ✅                                           | ❌                                 | 
| Progressive disclosure of content (single level of data)     | ❌                                           | ✅                                 | 
| Items includes arbitrary content and functionality           | ❌                                           | ✅                                 | 

> | Items includes supplementary actions                         | ✅                                           | ❌                                 | 

Reviewer: My read is that "supplementary actions" is a more limited subset of "arbitrary content and functionality".  So it seems strange that I can use an accordion for the broader set, but not for the more narrow set.

Phrasing as "Items **only** have supplementary actions" would make the distinction clearer.

Author: @angrycat9000 Some of the wording comes from prior research and can be somewhat imprecise. I've changed it to "associated actions" since supplementary isn't the best qualifier anyway. I also included "only" and fixed some conjugation typos.

```jsx
<Tree.Root>
  <Tree.Item label="Parent" aria-level={1} aria-posinset={1} aria-setsize={1} />
  <Tree.Item label="Child 1" aria-level={2} aria-posinset={1} aria-setsize={2} />
  <Tree.Item label="Child 2" aria-level={2} aria-posinset={2} aria-setsize={2}  />
</Tree.Root>
```

### Supplementary actions

Unlike the [**Accordion**](/components/accordion), items cannot take arbitrary content or functionality. The `Tree.Item` component itself is self-closed. However, the `actions` and `inlineActions` props lets you insert an array of supplementary actions. 

| Prop              | Action placement                                   |
| ----------------- | -------------------------------------------------- |
| `actions`         | Available in a dropdown menu in the `Tree.Item`    |
| `inlineActions`   | Available _inline_, outside of the dropdown menu   |

Each action must use a [`Tree.ItemAction`](/reference/structures/Tree/#Tree.ItemAction) component. Other components are not permitted.

```jsx
actions={[
  <Tree.ItemAction key="bring-to-front" label="Bring to front" />,
  <Tree.ItemAction key="send-to-back" label="Send to back" />

> ]}

Reviewer: Would include the context of wrapping this in `Tree.Item` to reinforce this is a prop on `Tree.Item`

May also want to include an `inlineActions` as an example as well.  These require an `icon` prop which `action` does not.

Author: @angrycat9000 I omitted the `Tree.Item` because it has a lot of required props (ARIA stuff mostly) so would become quite verbose. I think the preceding text and examples makes where to place this clear.

Reviewer: When I skimmed the page I skipped those parts and just went straight to the example.  Then ended up looking at the API docs to figure out of `Tree` or `Tree.Item` took the `action` prop.  

We could truncate the other props like below to still provide the context without losing the focus of the `action` prop.

```tsx
<Tree.Item
    actions={[
      <Tree.ItemAction key="bring-to-front" label="Bring to front" />,
      <Tree.ItemAction key="send-to-back" label="Send to back" />
    ]}
    {...otherProps}
/>

```

## Interaction

[`Tree.Item`s](/reference/structures/Tree/#Tree.Item) support two _primary_ interactions:

1. Expansion
2. Selection

### Expansion

> Any [`Tree.Item`](/reference/structures/Tree/#Tree.Item) can be expanded to reveal other `Tree.Item`s. If the `expandable` prop is omitted, the `Tree.Item` is considered a _leaf_. “Child 1” and “Child 2” are both _leaves_ in the following example.

Reviewer: Prop name typo
```suggestion
Any [`Tree.Item`](/reference/structures/Tree/#Tree.Item) can be expanded to reveal other `Tree.Item`s. If the `expanded` prop is omitted, the `Tree.Item` is considered a _leaf_. “Child 1” and “Child 2” are both _leaves_ in the following example.
```

To support full keyboard interaction, the left and right arrow keys are reserved for collapsing and expanding items respectively. The up and down arrows keys enable moving _between_ items.

:::

### Selection 

Item selection is facilitated with [`selected`](/reference/structures/Tree/#Tree.Item.selected) and the [`onSelectedChange`](/reference/structures/Tree/#Tree.Item.onSelectedChange) callback. No specific selection behavior is supported out of the box, since product needs diverge.

#### Single selection

One approach is to make items behave like radio buttons, wherein only one can be selected at a time.

::example{src="structures/Tree.select-one" min-width="300px" min-height="150px" vertical-stretch}

#### Cumulative selection

If desired, each [`Tree.Item`](/reference/structures/Tree/#Tree.Item) can act independently, as a simple toggle button. This means they can be selected cumulatively.

In the following example, “Item 1” and “Item 1.2” are selected from the outset. The “Select all” and “Deselect all” buttons cover bulk actions. 

::example{src="structures/Tree.select-many" min-width="300px" min-height="150px" vertical-stretch}

:::note[Disabling bulk actions]

> In the previous example, `useEffect` watches the selection object and disables the bulk actions where they're not applicable.

Reviewer: Let a comment about this in the code, may need to be updated with the code change

Author: https://github.com/iTwin/stratakit/pull/1639/commits/5f4325923095644dcaf44b964c2f6fe89c16de9a I removed the `useEffect` reference

:::

## Examples

> See the [Sandbox](/sandbox) (and the [Sandbox source code](https://github.com/iTwin/stratakit/blob/main/apps/test-app/app/sandbox/index.tsx)) for a working example with multiple features.

Reviewer: This ends up linking to `/docs/sandbox` which doesn't work.  Astro is doing some re-writing of links to make them relative to `/docs`

`../../../sandbox` works but is a bit weird.  Not sure if we have a better way to escape the base URL of `/docs`

Author: @mayank99 might be able to help with that.

Reviewer 2: We shouldn't be linking to the sandbox anyway. It's for internal testing and not meant to be a consumer-facing showcase.

Author: https://github.com/iTwin/stratakit/pull/1639/commits/30db45c2352304558f0e0a3e3fdaf4367445de36 I removed the whole **Example** section, since there are examples throughout.

---
title: Tree
description: Trees are interactive hierarchies.
status: stable
links:
  apiReference: /reference/structures/Tree
---

::example{src="structures/Tree.default" min-width="300px" min-height="150px" vertical-stretch}

The **Tree** component exposes application data as a hierarchy of nested, expandable levels. **Trees** can be explored and organized in different ways. They may offer searching, filtering, and sorting functionality.

> Like the [**Accordion**](/components/accordion), the **Tree** is useful for progressively disclosing content. Unlike the **Accordion**, this content can be nested to multiple levels of depth.

Reviewer 2: The last sentence is not accurate in saying "unlike the Accordion". There is nothing preventing accordions from being nested. The main difference is not the number of levels, it's the content itself.

Trees are for hierarchical datasets only, while accordions work for any arbitrary content. Crucially, a treeitem, unlike accordion content, is considered to be an interactive element and so it must not have nested actions (other than the built-in supplementary actions, which we have carefully designed).

Author: @mayank99 That you shouldn't nest Accordion items is longstanding UX advice, coming from the researchers. The Accordion doc carries that over:

> Don't nest **Accordions** inside one another.

But I think that's less relevant here, so I've just reworded regarding data versus content.

https://github.com/iTwin/stratakit/pull/1639/commits/c38a0089f63b1199767f077fbe9e0c5fe2eb6bd0

| Use case                                                     | [Tree](/components/tree)                     | [Accordion](/components/accordion) | 
| ------------------------------------------------------------ | -------------------------------------------- | ---------------------------------- | 
| Progressive disclosure of content (several levels of data)   | ✅                                           | ❌                                 | 
| Progressive disclosure of content (single level of data)     | ❌                                           | ✅                                 | 
| Items include arbitrary content and functionality            | ❌                                           | ✅                                 | 
| Items include associated actions only                        | ✅                                           | ❌                                 | 

## Structure

### Components

A **Tree** constitutes three key components:

- [`Tree.Root`](/reference/structures/Tree/#Tree.Root): The container, defining a single **Tree**.
- [`Tree.Item`](/reference/structures/Tree/#Tree.Item): Items belonging directly to `Tree.Root`.

> - [`Tree.ItemAction`](/reference/structures/Tree/#Tree.ItemAction): `Tree.Item` can be actionable when selected. `TreeItem.Action` defines a _supplementary_ action.

Reviewer 3: "actionable when selected" might get mixed with item selection (state).

Reviewer 2: ```suggestion
- [`Tree.ItemAction`](/reference/structures/Tree/#Tree.ItemAction): A [supplementary action](#supplementary-actions) for the `Tree.Item`.
```

### Hierarchy

> [`Tree.Item`s](/reference/structures/Tree/#Tree.Item) are supplied as children of [`Tree.Root`](/reference/structures/Tree/#Tree.Root). The components themselves are not nestable. Instead, the `aria-level` prop (required) describes the level of the item in the hierarchy. The [`aria-posinset`](/reference/structures/Tree/#Tree.Item.aria-posinset) and [`aria-setsize`](/reference/structures/Tree/#Tree.Item.aria-setsize) props are also required.

Reviewer 3: Link to a prop that is mentioned first time.

```suggestion
[`Tree.Item`s](/reference/structures/Tree/#Tree.Item) are supplied as children of [`Tree.Root`](/reference/structures/Tree/#Tree.Root). The components themselves are not nestable. Instead, the [`aria-level`](/reference/structures/Tree/#Tree.Item.aria-level) prop (required) describes the level of the item in the hierarchy. The [`aria-posinset`](/reference/structures/Tree/#Tree.Item.aria-posinset) and [`aria-setsize`](/reference/structures/Tree/#Tree.Item.aria-setsize) props are also required.
```

```jsx
<Tree.Root>
  <Tree.Item label="Parent" aria-level={1} aria-posinset={1} aria-setsize={1} />
  <Tree.Item label="Child 1" aria-level={2} aria-posinset={1} aria-setsize={2} />
  <Tree.Item label="Child 2" aria-level={2} aria-posinset={2} aria-setsize={2}  />
</Tree.Root>
```

### Supplementary actions

> Unlike the [**Accordion**](/components/accordion), items cannot take arbitrary content or functionality. The `Tree.Item` component itself is self-closed. However, the `actions` and `inlineActions` props lets you insert an array of supplementary actions. 

Reviewer 3: Link to props mentioned for 1st time

```suggestion
Unlike the [**Accordion**](/components/accordion), items cannot take arbitrary content or functionality. The `Tree.Item` component itself is self-closed. However, the [`actions`](/reference/structures/Tree/#Tree.Item.actions) and [`inlineActions`](/reference/structures/Tree/#Tree.Item.inlineActions) props lets you insert an array of supplementary actions. 
```

| Prop              | Action placement                                            |
| ----------------- | ----------------------------------------------------------- |
| `actions`         | Available in a dropdown menu                                |
| `inlineActions`   | Available _inline_, outside and preceding the dropdown menu |

Each action must use a [`Tree.ItemAction`](/reference/structures/Tree/#Tree.ItemAction) component. Other components are not permitted.

```jsx
actions={[
  <Tree.ItemAction key="bring-to-front" label="Bring to front" />,
  <Tree.ItemAction key="send-to-back" label="Send to back" />
]}
```

## Interaction

[`Tree.Item`s](/reference/structures/Tree/#Tree.Item) support two _primary_ interactions:

1. Expansion

> 2. Selection

Reviewer 3: ```suggestion
1. [Expansion](#expansion)
2. [Selection](#selection)
```

### Expansion

> Any [`Tree.Item`](/reference/structures/Tree/#Tree.Item) can be expanded to reveal other `Tree.Item`s. If the `expanded` prop is omitted, the `Tree.Item` is considered a _leaf_. “Child 1” and “Child 2” are both _leaves_ in the following example.

Reviewer 3: Need to mention (and link) to both `expanded` and `onExpandedChange` prop.

Author: https://github.com/iTwin/stratakit/pull/1639/commits/ba856a51729dd5ca95e3c2208669156db7df7f5d

```jsx
<Tree.Root>
  <Tree.Item 
    label="Parent" 
    aria-level={1} 
    aria-posinset={1} 
    aria-setsize={1} 
    expanded={isExpanded} 
    onExpandedChange={() => isExpanded = !isExpanded}
  />
  {isExpanded && (
    <Tree.Item label="Child 1" aria-level={2} aria-posinset={1} aria-setsize={2} />
    <Tree.Item label="Child 2" aria-level={2} aria-posinset={2} aria-setsize={2}  />
  )}
</Tree.Root>
```

A button to the left of each item is reserved for expanding and collapsing it. Clicking elsewhere on the item will perform [selection](#selection).

:::note[Keyboard behavior]

To support full keyboard interaction, the left and right arrow keys are reserved for collapsing and expanding items respectively. The up and down arrow keys enable moving _between_ items.

:::

### Selection 

> Item selection is facilitated with [`selected`](/reference/structures/Tree/#Tree.Item.selected) and the [`onSelectedChange`](/reference/structures/Tree/#Tree.Item.onSelectedChange) callback. No specific selection behavior is supported out of the box, since product needs diverge.

Reviewer 3: "since product needs diverge" can we reword this (bonus points for avoiding the word "product")

Author: I'm just adopting the Bentley lexicon 😉. I gave it a stab, for fun, though https://github.com/iTwin/stratakit/pull/1639/commits/19953163d8edb0b9325ca3a20e5e31590041b2af

#### Single selection

One approach is to make items behave like radio buttons, wherein only one can be selected at a time.

::example{src="structures/Tree.select-one" min-width="300px" min-height="150px" vertical-stretch}

#### Cumulative selection

If desired, each [`Tree.Item`](/reference/structures/Tree/#Tree.Item) can act independently, as a simple toggle button. This means they can be selected cumulatively.

In the following example, “Item 1” and “Item 1.2” are selected from the outset. The “Select all” and “Deselect all” buttons cover bulk actions. These are disabled when not applicable.

::example{src="structures/Tree.select-many" min-width="300px" min-height="150px" vertical-stretch}

> :::note[The `aria-multiselectable` attribute]

Reviewer 3: Linking once to https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-multiselectable would be nice

Author: https://github.com/iTwin/stratakit/pull/1639/commits/ba856a51729dd5ca95e3c2208669156db7df7f5d

> In the following example, “Item 1” and “Item 1.2” are selected from the outset. The “Select all” and “Deselect all” buttons cover bulk actions. These are disabled when not applicable.

Reviewer 2: There is no "Select all" or "Deselect all", as mentioned in https://github.com/iTwin/stratakit/pull/1639#issuecomment-5089014561.

```suggestion
In the following example, “Item 1” and “Item 1.2” are selected from the outset.
```

Selecting an item applies `aria-selected="true"`. Where it's possible to select multiple items, you must apply [`aria-multiselectable="true"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-multiselectable) to `Tree.Root`.

```jsx
<Tree.Root className={styles.tree} aria-multiselectable="true">
	...
</Tree.Root>
```

:::

## ✅ Do

- Do use **Tree** to let users explore multi-tiered data.

> - Do create expandable items that reveal nested items.

Reviewer 2: In other component docs, these items are not usually prefixed with "Do".

Reviewer 3: [Drop Do prefix](https://github.com/iTwin/stratakit/pull/1639/commits/0887bf401c1f204163d06cd64771b36a7ca0a00b)

- Do supply supplementary actions via the `actions` and `inlineActions` props.
- Do prioritize actions by adding them via `inlineActions`. 
- Do implement a selection behavior suited to your users.
- Do use an `aria-level` value that reflects the level of the item in the hierarchy. 

## ❌ Don't

- Don't make expandable items reveal items of the same `aria-level`. They must be the parent's level plus `1`.
- Don't try to place `Tree.Item`s inside `Tree.Item`s. Each `Tree.Item` is a sibling under a `Tree.Root` parent.
- Don't put any components besides [`Tree.ItemAction`](/reference/structures/Tree/#Tree.ItemAction) in [`Tree.Item`'s](/reference/structures/Tree/#Tree.Item) `action`.

> - Don't override the keyboard behaviors supplied.

Reviewer 3: Might be worth removing this. Instead we could mention that keyboard modifiers (ctrl/shift) can be implemented for multi-selection.

Author: https://github.com/iTwin/stratakit/pull/1639/commits/6517877c7fab3683ffb4718ab1914dbce3a37818 I've not mentioned the modifier keys, because I think that's a bigger topic. We could (should?) add another example that does that later. Instead, I just reworded to be more specific.

Reviewer 3: > - Don't remove the keyboard expansion and selection behaviors supplied.

Not a hard "Don't" rule, since:
- expand/select behavior is not really built in (consumers still need to hook it up and manage the state)
- more advanced selection management would currently require hooking up to lower level handlers (like `onClick`) e.g. https://github.com/iTwin/presentation/blob/91c6f0c89db1994cacb76a9a2ef82879c26d7e57/packages/hierarchies-react/src/presentation-hierarchies-react/stratakit/TreeRenderer.tsx#L378-L387
- is non expandable tree a valid use-case (Tree w/o expanders)?

Author: Hmm good point. I meant the accessibility provisions, but it's difficult to untangle, conceptually, from the rest. I've just removed the "don't" for now. https://github.com/iTwin/stratakit/pull/1639/commits/3380276922e4795988f0c7eb02bb89e674d503b9

Each action must use a [`Tree.ItemAction`](/reference/structures/Tree/#Tree.ItemAction) component. Other components are not permitted.

```jsx
actions={[
  <Tree.ItemAction key="bring-to-front" label="Bring to front" />,
  <Tree.ItemAction key="send-to-back" label="Send to back" />
]}
```

## Interaction

[`Tree.Item`s](/reference/structures/Tree/#Tree.Item) support two _primary_ interactions:

1. [Expansion](#expansion)
2. [Selection](#selection)

### Expansion

Any [`Tree.Item`](/reference/structures/Tree/#Tree.Item) can be expanded to reveal other `Tree.Item`s. If the [`expanded`](/reference/structures/Tree/#Tree.Item.expanded) prop is omitted, the `Tree.Item` is considered a _leaf_. “Child 1” and “Child 2” are both _leaves_ in the following example.

```jsx
<Tree.Root>
  <Tree.Item
    label="Parent"
    aria-level={1}
    aria-posinset={1}
    aria-setsize={1}
    expanded={isExpanded}
    onExpandedChange={() => isExpanded = !isExpanded}
  />
  {isExpanded && (
    <Tree.Item label="Child 1" aria-level={2} aria-posinset={1} aria-setsize={2} />
    <Tree.Item label="Child 2" aria-level={2} aria-posinset={2} aria-setsize={2}  />
  )}
</Tree.Root>
```

> A button to the left of each item is reserved for expanding and collapsing it, handled by [`onExpandedChange`](/reference/structures/Tree/#Tree.Item.onExpandedChange). Clicking elsewhere on the item will perform [selection](#selection).

Reviewer 2: > Clicking elsewhere on the item will perform selection.

This does not always apply.

Treeitems don't have to be selectable. If the `selected` prop is not passed, clicking on the treeitem will expand it (#1079). This can be seen in the [main demo](https://stratakit.bentley.com/1639/docs/components/tree/) at the top of the page.

Author: @mayank99 I've addressed this in https://github.com/iTwin/stratakit/pull/1639/commits/2a389ab38d9ee7294fe99a4e66b59868dc91298d. This on should be good to merge now? `ErrorRegion` should also be ready if we want to coordinate them.

#### Cumulative selection

If desired, each [`Tree.Item`](/reference/structures/Tree/#Tree.Item) can act independently, as a simple toggle button. This means they can be selected cumulatively.

In the following example, “Item 1.1” and “Item 3” are selected from the outset.

::example{src="structures/Tree.select-many" min-width="300px" min-height="150px" vertical-stretch}

:::note[The `aria-multiselectable` attribute]

Selecting an item applies `aria-selected="true"`. Where it's possible to select multiple items, you must apply [`aria-multiselectable="true"`](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Attributes/aria-multiselectable) to `Tree.Root`.

```jsx
<Tree.Root className={styles.tree} aria-multiselectable="true">
	...
</Tree.Root>
```

:::

## ✅ Do

- Use **Tree** to let users explore multi-tiered data.
- Create expandable items that reveal nested items.
- Supply supplementary actions via the `actions` and `inlineActions` props.
- Prioritize actions by adding them via `inlineActions`.
- Implement a selection behavior suited to your users.
- Use an `aria-level` value that reflects the level of the item in the hierarchy.

> ## ❌ Don't

Reviewer 2: ```suggestion
## 🚫 Don't
```
